src/handshake/client.rs

```diff
@@ -125,6 +129,14 @@ fn generate_request(request: Request, key: &str) -> Result<Vec<u8>> {
     .unwrap();
 
     for (k, v) in request.headers() {
+        if k == header::CONNECTION
+            || k == header::UPGRADE
+            || k == header::HOST
+            || k == header::SEC_WEBSOCKET_VERSION
+            || k == header::SEC_WEBSOCKET_KEY
+        {
+            continue;
+        }
```

Comment:
According to the RFC 6455:

```
   An unordered set of header fields comes after the leading line in
   both cases.  The meaning of these header fields is specified in
   Section 4 of this document.  Additional header fields may also be
   present, such as cookies [RFC6265]. 

   In compliance with [RFC2616], header fields in the handshake may be
   sent by the client in any order, so the order in which different
   header fields are received is not significant.
```

So I'm not sure if we're really prohibited to add additional headers even if they duplicate some of the previous headers provided that we always ensure that the unordered set of the required headers is always passed before user headers, so for the mandatory headers the standard values are going to be used, user values may then be used for other purposes (some custom user logic).

But even if we were to prohibit the user to specify some headers, I think we would better returning an error early instead of silently skipping them (may not be the expected behavior for the user).

Reply:
good suggestion, it's better to return a error.